src/axmysql/messagestore.go

```diff
@@ -0,0 +1,386 @@
+package axmysql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"time"
+
+	"github.com/jmalloc/ax/src/ax"
+	"github.com/jmalloc/ax/src/ax/persistence"
+)
+
+// MessageStore is an implementation of persistence.MessageStore that uses SQL
+// persistence.
+type MessageStore struct{}
+
+// AppendMessages appends one or more messages to a named stream.
+//
+// offset is a zero-based index into the stream. An error is returned if
+// offset is not the next unused offset in the stream.
+func (s MessageStore) AppendMessages(
+	ctx context.Context,
+	ptx persistence.Tx,
+	stream string,
+	offset uint64,
+	envs []ax.Envelope,
+) error {
+	tx := sqlTx(ptx)
+
+	var (
+		id  int64
+		err error
+	)
+
+	n := uint64(len(envs))
+
+	if offset == 0 {
+		id, err = s.insertStream(ctx, tx, stream, n)
+	} else {
+		id, err = s.incrStreamOffset(ctx, tx, stream, offset, n)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	g, err := s.incrGlobalOffset(ctx, tx, n)
+	if err != nil {
+		return err
+	}
+
+	for _, env := range envs {
+		if err := s.insertStreamMessage(
+			ctx,
+			tx,
+			g,
+			id,
+			offset,
+			env,
+		); err != nil {
+			return err
+		}
+
+		g++
+		offset++
+	}
+
+	return nil
+}
+
+// OpenStream opens a stream of messages for reading from a specific offset.
+func (MessageStore) OpenStream(
+	ctx context.Context,
+	ptx persistence.Tx,
+	stream string,
+	offset uint64,
+) (persistence.MessageStream, error) {
+	tx := sqlTx(ptx)
+
+	var id int64
+
+	if err := tx.QueryRowContext(
+		ctx,
+		`SELECT
+			stream_id
+		FROM messagestore_stream
+		WHERE name = ?`,
+		stream,
+	).Scan(
+		&id,
+	); err != nil {
+		return nil, err
+	}
+
+	return &MessageStream{
+		tx:     tx,
+		id:     id,
+		offset: offset,
+	}, nil
+}
+
+// insertStream inserts a new stream and returns its ID.
+// n is the number of events being appended.
+func (MessageStore) insertStream(
+	ctx context.Context,
+	tx *sql.Tx,
+	s string,
+	n uint64,
+) (int64, error) {
+	res, err := tx.ExecContext(
+		ctx,
+		`INSERT INTO messagestore_stream SET
+			name = ?,
+			next = ?`,
+		s,
+		n,
+	)
+	if err != nil {
+		return 0, err
+	}
+
+	return res.LastInsertId()
+}
+
+// incrStreamOffset increments the offset for the given stream by n.
+// It returns an error if o is not the next free offset.
+func (MessageStore) incrStreamOffset(
+	ctx context.Context,
+	tx *sql.Tx,
+	s string,
+	o uint64,
+	n uint64,
+) (int64, error) {
```

Comment:
Should this `int64` return value be documented?

Reply:
I don't think there's much need when there's only a single non-error return value.  The function description *does* document what it returns -- it's just not named.